src/Compilers/Core/Portable/Symbols/Attributes/CommonAttributeData.cs

```diff
@@ -292,7 +293,47 @@ private ObsoleteAttributeData DecodeObsoleteAttribute()
                 }
             }
 
-            return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError);
+            Debug.Assert(AttributeClass is object);
```

Comment:
>AttributeClass [](start = 25, length = 14)

An access to this property is going to allocate a public wrapper for this symbol and access to its members will do the same for them. Let's try delegating the ```IsProperty``` check to the derived language specific types, the should be able to do that by using internal symbols.

Reply:
Done

Reply:
>> An access to this property is going to allocate a public wrapper for this symbol and access to its members will do the same for them. Let's try delegating the IsProperty check to the derived language specific types, the should be able to do that by using internal symbols.

> Done

Still accessing the property.

---
In reply to: [398874477](https://github.com/dotnet/roslyn/pull/42518#discussion_r398874477) [](ancestors = 398874477)

Reply:
Ah yes. this assertion no longer has a purpose so will delete.

Reply:
Also there seem to be some usages of the property for MethodImpl and StructLayout attributes, perhaps we should refactor those as well.

Reply:
> Also there seem to be some usages of the property for MethodImpl and StructLayout attributes, perhaps we should refactor those as well. (in a future PR)

It looks like all those except one assert are for error reporting. That should be fine, formatting a symbol needs public symbol anyway.

---
In reply to: [399460296](https://github.com/dotnet/roslyn/pull/42518#discussion_r399460296) [](ancestors = 399460296)